clang/lib/CIR/Dialect/Transforms/TargetLowering/ItaniumCXXABI.cpp

```diff
@@ -67,6 +81,48 @@ bool ItaniumCXXABI::classifyReturnType(LowerFunctionInfo &FI) const {
   return false;
 }
 
+mlir::Type ItaniumCXXABI::getLLVMTypeForDataMember(
+    cir::DataMemberType type, mlir::TypeConverter &typeConverter) const {
+  // Itanium C++ ABI 2.3:
+  //   A pointer to data member is an offset from the base address of
+  //   the class object containing it, represented as a ptrdiff_t
+  const clang::TargetInfo &target = LM.getTarget();
+  unsigned width =
+      target.getTypeWidth(target.getPtrDiffType(clang::LangAS::Default));
+  return mlir::IntegerType::get(type.getContext(), width);
+}
+
+mlir::Attribute ItaniumCXXABI::lowerDataMemberConstant(
+    cir::DataMemberAttr attr, const mlir::DataLayout &layout,
+    const mlir::TypeConverter &typeConverter) const {
+  uint64_t memberOffset;
+  if (attr.isNullPtr()) {
+    // Itanium C++ ABI 2.3:
+    //   A NULL pointer is represented as -1.
+    memberOffset = -1ull;
+  } else {
+    // Itanium C++ ABI 2.3:
+    //   A pointer to data member is an offset from the base address of
+    //   the class object containing it, represented as a ptrdiff_t
+    auto memberIndex = attr.getMemberIndex().value();
+    memberOffset =
+        attr.getType().getClsTy().getElementOffset(layout, memberIndex);
+  }
+
+  auto underlyingIntTy = mlir::IntegerType::get(
+      attr.getContext(), layout.getTypeSizeInBits(attr.getType()));
+  return mlir::IntegerAttr::get(underlyingIntTy, memberOffset);
+}
+
+mlir::Operation *ItaniumCXXABI::lowerGetRuntimeMember(
+    cir::GetRuntimeMemberOp op, mlir::Type loweredResultTy,
+    mlir::Value loweredAddr, mlir::Value loweredMember,
+    mlir::OpBuilder &builder) const {
+  auto llvmElementTy = mlir::IntegerType::get(op.getContext(), 8);
+  return builder.create<mlir::LLVM::GEPOp>(
```

Comment:
We should not assume LLVM lowering at this level and emit a GEP here, target lowering is supposed to understand CIR to CIR. In theory CIR types and etc should be already on the right ABI, so that LLVM lowering has no extra work. Have you tried only returning the necessary information for lower to LLVM to emit the GEP?

Reply:
> In theory CIR types and etc should be already on the right ABI, so that LLVM lowering has no extra work.

Basically this PR does ABI lowering work. Thus I'm thinking that I should move the changes here to some prior pass that do ABI lowering works. LoweringPrepare seems good to me at first but I found it a bit difficult to lower type, attributes, and ops all at the same time there, and I'm not sure whether it's conceptually right to do ABI lowering during LoweringPrepare. Should we invent a new ABI lowering pass for this kind of stuff? CallConvLowering already does some ABI related lowering but that's for function calls and function signatures only.

Reply:
It's still fine if these things return CIR-only operations to LowerToLLVM, my understanding is that those will get scheduled to be lowered when the pass continues. Did you try that? Example: when lowering casts we call the rewriter to emit `cir::CmpOp`, etc.

Reply:
> Did you try that? Example: when lowering casts we call the rewriter to emit cir::CmpOp, etc.

Wow didn't try this. This is indeed a good way to resolve this if it works. Will try later.

Reply:
Updated.